File: clients/swift/AgentHostProtocol/Sources/AgentHostProtocolClient/Hosts/HostState.swift
// HostState — connection state for a single host.

import Foundation

/// Connection state for a single host.
public enum HostState: Sendable, Equatable {
    /// The host has been added but no transport is open.
    case disconnected
    /// A transport is being opened or the `initialize` handshake is in flight.
    case connecting
    /// The host is fully connected and serving subscriptions.
    case connected
    /// A previous connection dropped; the supervisor is retrying with backoff.
    ///
    /// `attempt` is one-based and resets after a successful connect when the
    /// host's `ReconnectPolicy.resetOnSuccess` is `true`.
    case reconnecting(attempt: Int)
    /// Reconnect attempts were exhausted (or `ReconnectPolicy.disabled` was
    /// configured) and the host is no longer trying. The supervisor still
    /// services `snapshot`, manual `reconnect`, and `shutdown` commands while
    /// in this state.
    case failed(reason: String)

    /// Convenience: is the host currently `.connected`?
    public var isConnected: Bool {
        if case .connected = self { return true }
        return false
    }

    /// Convenience: is the host in a terminal failure state?
    public var isFailed: Bool {
        if case .failed = self { return true }
        return false
    }
}

File: clients/dotnet/src/AgentHostProtocol/Hosts/HostState.cs
// Host lifecycle state.
#nullable enable

using System;

namespace Microsoft.AgentHostProtocol.Hosts;

/// <summary>Lifecycle states a host can be in.</summary>
public enum HostStateKind
{
    /// <summary>Added but no transport is open.</summary>
    Disconnected,
    /// <summary>Transport is being opened or <c>initialize</c> is in flight.</summary>
    Connecting,
    /// <summary>Fully connected and serving subscriptions.</summary>
    Connected,
    /// <summary>Previous connection dropped; supervisor is retrying.</summary>
    Reconnecting,
    /// <summary>Reconnect attempts exhausted (or disabled).</summary>
    Failed,
}

/// <summary>Current lifecycle state of a host.</summary>
public sealed class HostState
{
    /// <summary>The state kind.</summary>
    public HostStateKind Kind { get; init; }

    /// <summary>Consecutive reconnect attempt counter.</summary>
    public uint Attempt { get; init; }

    /// <summary>The error that put the host into its current state, if any.</summary>
    public Exception? Error { get; init; }

    /// <inheritdoc />
    public override string ToString() => Kind switch
    {
        HostStateKind.Disconnected => "disconnected",
        HostStateKind.Connecting => "connecting",
        HostStateKind.Connected => "connected",
        HostStateKind.Reconnecting => "reconnecting",
        HostStateKind.Failed => "failed",
        _ => "unknown",
    };
}
